### DISCUSSION TITLE
fly login command hangs indefinitely with port forwarding session following upgrade from v4.2.3 -> v6.3.0

### DISCUSSION THREAD

**Discussion by tmateen-tw** [concourse/concourse#5827] [Help & Support]
# Summary
The `fly login` command hangs indefinitely when used with a Concourse v6.3.0 instance running in a private subnet while using port forwarding to communicate with the instance. Specifically, the [retrieving UserInfo](https://github.com/concourse/concourse/blob/ac5456701113a415b27b657ba23d95229ea598a1/fly/commands/login.go#L151) step ( i.e., `/api/v1/user` call) in the login flow hangs. This seems to be because concurrent connections are not allowed with port forwarding sessions - but the login command succeeded with the exact same set up when used with Concourse/fly v4.2.3. Removing the `/api/v1/user` step of the fly login code makes the command succeed.
# Steps to reproduce
1.  **Infrastructure Setup:**
    *   We have a Concourse v6.3.0 Web instance running with [AWS SSM Port Forwarding session](https://docs.aws.amazon.com/systems-manager/latest/userguide/session-manager-working-with-sessions-start.html#sessions-start-port-forwarding) for `fly` to be able to communicate with the concourse instance.
1. Running the `fly login` command with the proper credentials.
# Expected results
Login succeeds with `target saved`.
# Actual results
The login command hangs indefinitely at the [retrieving UserInfo](https://github.com/concourse/concourse/blob/master/fly/commands/login.go#L151) step.
# Additional Context
We found that commenting out the [retrieving UserInfo](https://github.com/concourse/concourse/blob/master/fly/commands/login.go#L151) step in the fly login command leads to the command executing succesfully. 
Further research using the `fly curl` command with the `/api/v1/user` endpoint lead to the command succeeding if the [target.Validate()](https://github.com/concourse/concourse/blob/master/fly/commands/curl.go#L27) command is commented out before, as to not have two API calls. 
A similiar discussionon in the [AWS Forums](https://forums.aws.amazon.com/thread.jspa?messageID=927704) also seems to lead to the notion that concurrent connections are not supported with port forwarding.
We would like to ask whether it would be possible to close request connections after the validation calls succeed/fail? As this issue was not there (perhaps due to multiple calls not being made in the same command) in Fly v4.2.3.

**Comment by jamieklassen** [comment]
First off, good job with the debugging!

I'm not sure this is actionable yet, so I'm converting it to a discussion following https://blog.concourse-ci.org/nobody-wants-a-stale-bot/#an-update-on-triage. Can you mention what auth connector you are using (i.e. GitHub, LDAP, CF auth, etc)? You didn't mention anything in particular so I assume you are using a local user and running `fly login` with the `-u` and `-p` flags. Is this correct?

**Reply by tmateen-tw** [reply]
Hi,

Thanks for the feedback! Yes that's correct, we are using a local user to login in this case.

If it makes sense to mention, we also have OIDC Auth set up alongside, but we **always** use the local user to login through a PortForwarding session instead because our the web instance has an extra layer of auth configured with its loadbalancer.

**Reply by tmateen-tw** [reply]
Hi @jamieklassen,

any thoughts/suggestions of possible directions we can head in for this?

**Reply by jamieklassen** [reply]
This actually sounds like some pretty low-level Go stuff. Reading [the docs](https://golang.org/pkg/net/http/#Transport), I see

> By default, Transport caches connections for future re-use. This may leave many open connections when accessing many hosts. This behavior can be managed using Transport's CloseIdleConnections method and the MaxIdleConnsPerHost and DisableKeepAlives fields. 

which has me wondering whether patching this

https://github.com/concourse/concourse/blob/a799028ee9248c00ddd1e3696feb43ee0d9228a3/fly/rc/target.go#L471-L480

to include `DisableKeepAlives: true` might help. I tried it out locally and it seemed to work and passed the tests, but I haven't been able to reproduce your AWS SSM port forwarding issue. I tried various concurrency-limiting reverse proxies but couldn't find one with an equivalent limitation 🤷‍♀️ . The only use case I can imagine this breaking is when you are using `fly watch` to stream build logs and there is a long delay where a build has no output, disabling keepalives might cause your connection to time out?

Another approach that might help would be for `client.GetInfo` here:

https://github.com/concourse/concourse/blob/a799028ee9248c00ddd1e3696feb43ee0d9228a3/go-concourse/concourse/info.go#L11-L15

 to have some way of calling `CloseIdleConnections` on the `http.Client` inside `client.connection` (yes, `target` has a `client` that has a `connection` that has a `client` whose `transport` caches 'connections' -- naming is very hard) -- it might take exposing another method on `Connection` to do it, though.

Does this help? It sounds like you're in a position to try out patching/building/testing with fly in your setup. I'm a bit puzzled about how best to test this

**Reply by tmateen-tw** [reply]
Hi @jamieklassen , 

Sorry for getting back to you so late, but thanks a lot for such a detailed response! We tried it out with the flag you suggested to add into the fly basic auth transport http client, and it works perfectly. Specifically:

```patch
diff --git a/fly/rc/target.go b/fly/rc/target.go
index 959db56e6..23717c514 100644
--- a/fly/rc/target.go
+++ b/fly/rc/target.go
@@ -593,6 +593,7 @@ func transport(insecure bool, caCertPool *x509.CertPool, clientCertificate []tls
                        Timeout: 10 * time.Second,
                }).Dial,
                Proxy: http.ProxyFromEnvironment,
+               DisableKeepAlives: true,
        }
 
        return transport
```

Would it make sense to have this option enabled by default/ as a configuration option?

Thanks again for the fast feedback!

**Reply by jamieklassen** [reply]
Cool, glad my research amounted to something!

Disabling keepalives in the fly client seems reasonable - I imagine the caching of connections is mainly useful for cases when you're making a ton of requests, and fly (as a generally-human-interaction-bound CLI) never has any explosive number of requests. It might be OK to set this default always and not even make it configurable, but my only concern would be if you `fly watch`ed a build where there were was no output for a long time, would the command timeout and fail? Maybe you could raise a PR for this and we could get a few more sets of eyes on the idea?
